src/backend_telemetry_buffer.rs

```diff
@@ -0,0 +1,169 @@
+use pgx::*;
+
+#[pg_schema]
+mod _prom_ext {
+    use pgx::{prelude::*, *};
+    use std::{
+        mem::MaybeUninit,
+        sync::{Mutex, Once},
+    };
+
+    extension_sql!(
+        r#"
+        CREATE TYPE _prom_ext.backend_telemetry_rec AS (
+            time timestamp with time zone,
+            value BIGINT
+        );
+        "#,
+        name = "backend_telemetry_rec_decl"
+    );
+
+    const COMPOSITE_TYPE_NAME: &str = "_prom_ext.backend_telemetry_rec";
+
+    type BufferItem = composite_type!(COMPOSITE_TYPE_NAME);
+    struct Buffer {
+        mem_ctx: PgMemoryContexts,
+        list: PgList<BufferItem>,
+    }
+
+    struct BufferIter<'b> {
+        buffer: &'b mut Buffer,
+        pos: usize,
+    }
+
+    impl<'a> Iterator for BufferIter<'a> {
+        type Item = BufferItem;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            // We are not deallocaing anything here because Drop does it wholesale.
+            let item = self.buffer.list.get_ptr(self.pos);
+            self.pos += 1;
+            // SAFETY: push_rec is expected to be the only funciton appending to buffer.list
+            item.map(|i| unsafe {
+                PgHeapTuple::from_datum_in_memory_context(
+                    PgMemoryContexts::CurrentMemoryContext,
+                    i.into(),
+                    false,
+                    BufferItem::type_oid(),
+                )
+                .unwrap()
+            })
+        }
+
+        fn size_hint(&self) -> (usize, Option<usize>) {
+            let remaining = self.buffer.list.len() - self.pos;
+            (remaining, Some(remaining))
+        }
+    }
+
+    impl<'a> Drop for BufferIter<'a> {
+        fn drop(&mut self) {
+            // Drop on the list calls list_free(). For it to work without segfaults we first 
+            // need to replace the list (triggering the Drop) and only then call mem_ctx.reset()
+            self.buffer.list = PgMemoryContexts::CacheMemoryContext.switch_to(|_| PgList::new());
```

Comment:
can't this be an assignment to NIL instead?

Reply:
`PgList::new()` is the `NIL`